File: 2019/day04-2.hs
import Data.List (group)
import Data.Char (ord)

solution :: String -> Int
solution content = length . filter (increasing . show) . filter (twoAdjacentEqual . show) $ [fst..lst]
    where
        limits = parse content
        fst = head limits
        lst = last limits

twoAdjacentEqual :: String -> Bool
twoAdjacentEqual xs = any (==2) (map length . group $ xs)

increasing :: String -> Bool
increasing xs = all (\(a,b) -> a == b || ord a < ord b) (zip xs (tail xs))

parse :: String -> [Int]
parse = map read . splitOn '-'

splitOn :: Char -> String -> [String]
splitOn c xs = helper [""] c xs
  where
    helper o c "" = o
    helper o c (x : "")
      | x == c = o
      | otherwise = init o ++ [last o ++ [x]]
    helper o c (x : xs)
      | x == c = helper (o ++ [""]) c xs
      | otherwise = helper (init o ++ [last o ++ [x]]) c xs

main :: IO ()
main = do readFile "day04.dat" >>= print . solution

File: 2019/day04_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from itertools import groupby
from typing import Any


def solution(content: str) -> Any:
    lo, hi = [int(x) for x in content.split("-")]
    return len(
        [
            number
            for number in (str(x) for x in range(lo, hi))
            if all(a <= b for a, b in zip(number, number[1:]))
            and any(len(list(g)) == 2 for _, g in groupby(number))
        ]
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
